src/libmongoc/src/mongoc/mongoc-client-session.c

```diff
@@ -639,8 +668,28 @@ _mongoc_client_session_handle_reply (mongoc_client_session_t *session,
          mongoc_client_session_advance_cluster_time (session, &cluster_time);
       } else if (!strcmp (bson_iter_key (&iter), "operationTime") &&
                  BSON_ITER_HOLDS_TIMESTAMP (&iter) && is_acknowledged) {
-         bson_iter_timestamp (&iter, &t, &i);
-         mongoc_client_session_advance_operation_time (session, t, i);
+         bson_iter_timestamp (&iter, &operation_t, &operation_i);
+         mongoc_client_session_advance_operation_time (
+            session, operation_t, operation_i);
+      } else if (!strcmp (bson_iter_key (&iter), "cursor") &&
+                 mongoc_session_opts_get_snapshot (&session->opts) &&
+                 !session->snapshot_time_set) {
+         /* If cursor is present, snapshot is enabled for the session,
+          * and snapshot_time has not already been set, try to find
+          * atClusterTime in cursor field to set snapshot_time_timestamp and
+          * increment. */
+         bson_iter_recurse (&iter, &cursor_iter);
+
+         while (bson_iter_next (&cursor_iter)) {
+            /* If atClusterTime is in cursor and is a valid timestamp, use it to
+             * set snapshot_time_timestamp and increment. */
+            if (!strcmp (bson_iter_key (&cursor_iter), "atClusterTime") &&
+                BSON_ITER_HOLDS_TIMESTAMP (&cursor_iter)) {
+               bson_iter_timestamp (&cursor_iter, &snapshot_t, &snapshot_i);
+               _mongoc_client_session_set_snapshot_time (
+                  session, snapshot_t, snapshot_i);
+            }
+         }
```

Comment:
I think this should also inspect for a top-level `atClusterTime` in the server reply (not just within `cursor`).

This would support the use-case where a user runs `distinct` with a generic command helper.

Reply:
Added. Didn't have this before since I was only appending in `_mongoc_cursor_new_with_opts`, so the use-case of `distinct` with a generic command helper was not supported.